<a href="https://colab.research.google.com/github/nrannisak2/2026_Intro_Python/blob/main/Exercise_1_2_data_structures_and_control_flow_exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

> Notebook-friendly copy of `part-I/1.2-data-structures-and-control-flow-exercises.ipynb`, generated by `tools/make_live.py`. Edit the book notebook, not this file.

In [70]:
# --- environment setup (generated, not part of the lesson) ---
# Colab and Kaggle do not ship every package this notebook imports.
# This is a no-op in an environment that is already set up.
import importlib.util
import subprocess
import sys

for module, package in {"pooch": "pooch"}.items():
    if importlib.util.find_spec(module) is None:
        subprocess.run([sys.executable, "-m", "pip", "install", "-q", package], check=True)

# Exercises

**ℹ️ How to use these exercises**

Attempt each from scratch in the empty code cell. Write plain functions with clear names, and avoid a mutable default argument.

## Exercise 1: Lists — build, modify, slice

Start from the daily mean temperatures `[-2.3, -1.1, 0.4, 1.2, -0.8]` (°C).

1. Append a sixth reading, `-3.1`.
2. Remove the value `0.4` (the sensor flagged it as unreliable).
3. Print the first reading, the last reading, and a slice containing the second and third.
4. Print how many readings remain.

In [71]:
# Your solution here
daily_temp_c = [-2.3, -1.1, 0.4, 1.2, -0.8]
daily_temp_c.append(-3.1)
daily_temp_c.remove(0.4)

print(daily_temp_c[0], daily_temp_c[-1], daily_temp_c[1:3])
print(len(daily_temp_c))

-2.3 -3.1 [-1.1, 1.2]
5


## Exercise 2: Dictionaries — build, look up, remove

You are given station records as a list of tuples `(code, name, elevation_m)`:

```python
[("JFJ", "Jungfraujoch", 3571), ("BAS", "Basel-Binningen", 316), ("LUG", "Lugano", 273)]
```

1. Build a dict mapping each `code` to its `elevation_m`.
2. Print the elevation for `"JFJ"`.
3. Look up `"ZRH"` with `.get`, returning `"unknown"` if it is absent.
4. Delete the entry for `"LUG"`, then print each remaining code and its elevation using
   `.items()`.

In [72]:
# Your solution here
# Tuples one
records = [
    ("JFJ", "Jungfraujoch", 3571),
     ("BAS", "Basel-Binningen", 316),
      ("LUG", "Lugano", 273)
      ]

# 1. Build a dict mapping each code to a dict containing its name and elevation_m dictionary within a dictionary --> one that we use
stations = {}
for st_code, st_name, elevation in records:
    stations[st_code] = {"st_name": st_name, "elevation": elevation}

# 2. Print the elevation for "JFJ"
print(f"JFJ elevation: {stations["JFJ"]["elevation"]}")

# 3. Look up "ZRH" with .get, returning "unknown" if it is absent.
print(stations.get("ZRH", "unknown"))

# 4. Delete the entry for "LUG", then print each remaining code and its elevation using .items().
del stations["LUG"]

for code, details in stations.items():
    print(f"{code}: {details['elevation']}")

JFJ elevation: 3571
unknown
JFJ: 3571
BAS: 316


## Exercise 3: A pure function

Write a function `mean(values)` (no type hints) that returns the average of a list of numbers. Test it on `[6.1, 7.4, 5.9, 8.2, 6.8]`, printing the result rounded to two decimals.

In [73]:
# Your solution here
# Try out mean(values) with plain function
values = [6.1, 7.4, 5.9, 8.2, 6.8]
print(f'mean values: {mean(values)}')

# With function
# Using for & return
def mean(total):
    """Return the the mean of a list of numbers."""
    total = 0.0
    for v in values:
        total += v
    return total / len(values)
print(f'mean values: {mean(values):.2f}')

# Using pure function
def mean(values):
    return sum(values) / len(values)

print(f"mean values: {mean(values):.2f}")

mean values: 6.88
mean values: 6.88
mean values: 6.88


## Exercise 4: A function with a default argument

Write a function `is_warm(temp_celsius, threshold=10.0)` (no type hints) that returns `True` when the temperature is at or above the threshold and `False` otherwise, using `if`/`else`. Test it on 6.1 °C with the default threshold, and again on 6.1 °C with a threshold of 5.0 °C passed by keyword.

In [112]:
# Your solution here
def is_warm(temp_c, threshold=10):
  if temp_c >= threshold:
    return True
  else:
    return False

print("Is the temperature 6.1 °C warm?", is_warm(6.1))
print(f'If we change the threshold to 5.0 °C, is the temperature warm? {is_warm(6.1, 5.0)}')

Is the temperature 6.1 °C warm? False
If we change the threshold to 5.0 °C, is the temperature warm? True


## Exercise 5: Pair two stations with zip

Two stations recorded the same five days:

```python
dates = ["Mon", "Tue", "Wed", "Thu", "Fri"]
jfj_celsius = [-2.3, -1.1, 0.4, 1.2, -0.8]
bas_celsius = [6.1, 7.4, 5.9, 8.2, 6.8]
```

Using `zip`, print one line per day showing the date and the temperature difference
(Basel minus Jungfraujoch), rounded to two decimals.

In [75]:
# Your solution here
dates = ["Mon", "Tue", "Wed", "Thu", "Fri"]
jfj_celsius = [-2.3, -1.1, 0.4, 1.2, -0.8]
bas_celsius = [6.1, 7.4, 5.9, 8.2, 6.8]

for date, jfj_temp, bas_temp in zip(dates, jfj_celsius, bas_celsius):
  print(date, f"{jfj_temp-bas_temp:.2f}")


Mon -8.40
Tue -8.50
Wed -5.50
Thu -7.00
Fri -7.60


## Exercise 6: Number the readings with enumerate

A station wrote one reading per day for a week:

```python
week_celsius = [6.1, 7.4, 5.9, 8.2, 11.3, 9.8, 6.8]
```

1. Using `enumerate`, print one line per reading: the day number and the temperature. Count days
   from 1, not 0 — `enumerate` starts at 0, so the day number is the index plus one.
2. In the same loop, keep track of the warmest reading seen so far and the day it fell on, in two
   variables defined before the loop. Do not use `max`.
3. After the loop, print the warmest reading and its day number.

In [76]:
# Your solution here
# Defining variables
week_celsius = [6.1, 7.4, 5.9, 8.2, 11.3, 9.8, 6.8]
warmest_temp = week_celsius[0]
warmest_day = 1

# Using enumerate
print("Day Temperature")
for i, temp in enumerate(week_celsius):
  day_number = i+1
  print(f"Day {day_number}: {temp} °C")

# Keep tract on warmest reading seen so far
  if temp > warmest_temp:
      warmest_temp = temp
      warmest_day = day_number

print(f"The warmest reading so far was: {warmest_temp} Celsius on day {warmest_day}")


Day Temperature
Day 1: 6.1 °C
Day 2: 7.4 °C
Day 3: 5.9 °C
Day 4: 8.2 °C
Day 5: 11.3 °C
Day 6: 9.8 °C
Day 7: 6.8 °C
The warmest reading so far was: 11.3 Celsius on day 5


## Exercise 7: while, break, and continue

A logger returned six readings, two of which are missing:

```python
mixed_celsius = [6.1, None, 5.9, None, 9.5, 6.8]
```

Using a `while` loop over the indices:

1. Skip any missing value with `continue`.
2. Stop at the first reading above 9.0 °C with `break`, storing it in `first_hot`.
3. Print `first_hot` after the loop. Define it before the loop so the code is safe even if no
   reading exceeds the threshold.

In [77]:
# Your solution here
mixed_celsius = [6.1, None, 5.9, None, 9.5, 6.8]
first_hot = None
i = 0

while i < len(mixed_celsius):
  temp = mixed_celsius[i]

  if temp is None:
    i += 1
    continue

  if temp > 9.0:
    first_hot = temp
    break

  i += 1

print(f"Found a hot day, stopping! {first_hot}")

# Hint: continue jumps straight back to the while line, skipping everything below it in the
# loop body, including an index update at the bottom of the loop

Found a hot day, stopping! 9.5


## Exercise 8: Default arguments and returning two values

Write a function `summarise(values, scale=1.0)` that multiplies every value by `scale` and
returns **two** things: the total and the mean.

1. Call it on `[6.1, 7.4, 5.9, 8.2, 6.8]` with the default scale, unpacking the result into
   `total_celsius` and `mean_celsius`.
2. Call it again with `scale=2.0` passed by keyword, and print only the mean.
3. Call it once more and store the result in a single name. Print its `type` — what did you get?

In [78]:
# Your solution here

# 1
def summarise(value, scale=1.0):
  """Return the total and the mean of a list of numbers."""
  total = sum(value)*scale
  mean = total / len(value)
  return total, mean
# unpack
total_celsius, mean_celsius = summarise([ 6.1, 7.4, 5.9, 8.2, 6.8])
print(total_celsius, mean_celsius)

# 2
# call with scale=2.0
def summarise(value, scale=2.0):
  """Return the total and the mean of a list of numbers."""
  total = sum(value)*scale
  mean = total / len(value)
  return total, mean
total_celsius, mean_celsius = summarise([ 6.1, 7.4, 5.9, 8.2, 6.8])
print(mean_celsius)

# 3
# Call it once more and store the result in a single name
summarise_result = total_celsius, mean_celsius
print(summarise_result)
print(type(summarise_result))

34.4 6.88
13.76
(68.8, 13.76)
<class 'tuple'>


## Exercise 9: A variable number of arguments

Write a function `warmest(*temps_celsius)` (no type hints) that returns the highest value among
any number of positional arguments, using a loop and a running maximum (not `max`). Test it on
three individual readings passed directly, and again by unpacking the list
`[6.1, 7.4, 5.9, 8.2, 6.8]` with `*`.

In [79]:
# Your solution here

def warmest(*temps_celsius): # collect any number of positional arguments into a tuple
  warmest_temp = temps_celsius[0]

  for temp in temps_celsius:
    if temp > warmest_temp:
      warmest_temp = temp
  return warmest_temp

# testing - manual testing
result1 = warmest(6.1, 7.4, 5.9)
print(f"warmest of 3: {result1} °C")

# testing 2 - unpacking list using *
week_celsius = [6.1, 7.4, 5.9, 8.2, 6.8]
result2 = warmest(*week_celsius) # spread this iterable out into separate positional arguments
print(f"warmest of 5: {result2} °C")

warmest of 3: 7.4 °C
warmest of 5: 8.2 °C


## Exercise 10: Fix the shared-default bug

An assistant wrote this function; it is correct the first time but wrong on later calls:

```python
def collect_above(values, threshold, out=[]):
    for v in values:
        if v > threshold:
            out.append(v)
    return out
```

In one comment, explain the bug, then rewrite it so each call starts fresh. Demonstrate with two
separate calls, assigning each result to its own variable.

In [80]:
# Your solution here

# initial code
def collect_above(values, threshold, out=[]):
    for v in values:
        if v > threshold:
            out.append(v)
    return out
# the bug
# example
result1 = collect_above([6.1, 7.4, 5.9], 6.0)
print(result1)
result2 = collect_above([6.1, 7.4, 5.9, 8.2, 6.8], 6.5)
print(result2)
# result1 get carried away to result2 because
# 1. out=[] is mutable, and out.append(v) mutates the same list
# moving out=[] into the body means each call can start fresh! :)

# rewrite the code
def collect_above(values, threshold, out=None):
  if out is None:
    out = []
  for v in values:
    if v > threshold:
      out.append(v)
  return out
# reprint the example
result1 = collect_above([6.1, 7.4, 5.9], 6.0)
print(result1)
result2 = collect_above([6.1, 7.4, 5.9, 8.2, 6.8], 6.5)
print(result2)

[6.1, 7.4]
[6.1, 7.4, 7.4, 8.2, 6.8]
[6.1, 7.4]
[7.4, 8.2, 6.8]


## Going deeper (optional)

### Exercise 11: Document your function with a docstring

Take the `mean(values)` you wrote in Exercise 3 and give it a docstring — a one-line summary in
triple quotes on the first line of the body. State what it returns *and* the unit it assumes, so
the function carries information the name alone cannot.

Then print `mean.__doc__`, and call `help(mean)` to see what a reader of your code would see.

In [81]:
# Your solution here



### Exercise 12: Tuples — unpacking and immutability

A station's position is stored as a tuple, `station_coords = (46.5475, 7.9853)`.

1. Unpack it into `lat` and `lon` in one line, and print both.
2. Try to change the latitude to `47.0`. Catch the resulting error with `try`/`except TypeError`
   and print the message.
3. In a comment, say why a coordinate pair is better stored as a tuple than as a list.

In [82]:
# Your solution here

### Exercise 13: Means and a filter, with comprehensions

Recreate a `stations` dict where each code maps to a dict containing `readings_celsius`:

```python
{"JFJ": {"readings_celsius": [-2.3, -1.1, 0.4, 1.2, -0.8]},
 "BAS": {"readings_celsius": [6.1, 7.4, 5.9, 8.2, 6.8]},
 "LUG": {"readings_celsius": [9.4, 10.2, 8.8, 11.1, 9.9]}}
```

With a dict comprehension, build `means = {code: mean_temp}`, calling the `mean` function you
wrote in Exercise 3 rather than summing inline. Then, with a list comprehension, build the list
of codes whose mean is below 0 °C.

In [83]:
# Your solution here

### Exercise 14: Spot the impure function

This function is meant to return the readings without the last one:

```python
def drop_last(readings):
    readings.pop()
    return readings
```

1. Call it on `original = [6.1, 7.4, 5.9]`, then print `original`. What happened?
2. Rewrite it as a pure function that leaves the caller's list untouched, and demonstrate that
   `original` is unchanged after the call.
3. In one comment, connect this to the mutable-default bug from the lecture.

In [84]:
# Your solution here
def drop_last(readings):
    readings.pop()
    return readings

original = [6.1, 7.4, 5.9]
result = drop_last(original)
print("result", result)
print("original:", original)


result [6.1, 7.4]
original: [6.1, 7.4]


### Exercise 15: Reject impossible temperatures

Write `to_kelvin(temp_celsius)` that raises a `ValueError` when the input is below absolute zero
(−273.15 °C) and otherwise returns the temperature in kelvin. Give it a docstring and a type hint.

Loop over `[20.0, -300.0, 0.0]`, calling it inside a `try`/`except` so that the invalid value is
reported without stopping the loop.

In [85]:
# Your solution here

## Exercise 16: A year of station data, continued

The optional section ends with Exercise 15. Exercises 16 and 17 are the main exercises of this
subchapter.

In subchapter 1.1 you fetched this file, worked out its structure, and converted a single
record to °C. You did not yet have the tools to handle all of it at once — now you do.

The goal: turn a year of daily maximum temperatures into a monthly summary, and write that
summary out as a new file. The input file has the same three columns as in Exercise 9 of the
previous subchapter: the station code, the day as `dd.mm.yy`, and the daily maximum temperature
in degrees Fahrenheit. Your monthly summary will have its own columns: `month`, `n_obs`
(the number of observations), and `mean_temp_celsius`.

In [86]:
# Pre-supplied: download the data file and cache it locally.
# You do not need to understand this cell yet — fetching data is covered in the
# reproducible-data-pipelines bonus subchapter.
import pooch

data_file = pooch.retrieve(
    url="https://raw.githubusercontent.com/gse-unil/2026_MLEES_book/main/data/part-I/station_iib_daily_max_temp_2022.csv",
    known_hash="sha256:034755fb289b4e157e5d029995481786a359eb25a80df62c09ee83d063013144",
    fname="station_iib_daily_max_temp_2022.csv",
    path=pooch.os_cache("mlees"),
)
print("data cached at:", data_file)

data cached at: /root/.cache/mlees/station_iib_daily_max_temp_2022.csv


**Step 1.** Write a function `fahrenheit_to_celsius(temp_fahrenheit)` that returns the
temperature in °C. Test it on the freezing point of water (32 °F should give 0.0 °C) and on
212 °F. A one-line docstring is optional here (see Exercise 11).

You wrote this conversion inline in 1.1. Steps 2 and 4 both need it again; with a function,
they call the same tested code instead of repeating the arithmetic.

In [87]:
def fahrenheit_to_celsius(temp_fahrenheit):
  """Return temperature in celsius"""
  return (temp_fahrenheit - 32) * 5 / 9

freezing_point = fahrenheit_to_celsius(32)
print(freezing_point)

print(fahrenheit_to_celsius(212))


# Your solution here
# Hint: °C = (°F - 32) * 5 / 9
# Hint (optional, see Exercise 11): a docstring goes on the first line of the function body,
# in triple quotes

0.0
100.0


**Step 2.** Open the file, read all lines, and loop over the records (skipping the header).
For each one, extract the month and the temperature in °C. Print the first three as a check.

Sixteen days in this file have an empty temperature field — a real gap in the station's record,
late October to mid-November. `float("")` raises `ValueError`, so your loop has to decide what to
do with those days before it can get to the end of the file. Skip them, and count how many you
skipped. After the loop, print how many records you converted and how many you skipped; the two
should add up to 365.

In [88]:
# Your solution here
# Hint: open it as in 1.1, with Path(data_file).open("r", encoding="utf-8") in a with block
# Hint: lines[1:] skips the header row

from pathlib import Path

data_path = Path(data_file)

with data_path.open("r", encoding="utf-8") as fhandler:
  raw_readlines = fhandler.readlines()

print(raw_readlines[1:])

# Hint: .strip().split(",") gives you the three fields

fields = raw_readlines[1].strip().split(",")
print(fields)

# Hint: the month is the middle piece of the day field — split it on "."

date = fields[1]
new_date = date.split(".")
print(new_date)

month = new_date[1]
print(month)

# Hint: use a counter that you update with += inside the loop

# Hint: a day with no reading has fields[2] == "" — `continue` moves on to the next line

['IIB,01.01.22,14.7\n', 'IIB,02.01.22,8.6\n', 'IIB,03.01.22,24.4\n', 'IIB,04.01.22,33.8\n', 'IIB,05.01.22,21.6\n', 'IIB,06.01.22,1.4\n', 'IIB,07.01.22,8.6\n', 'IIB,08.01.22,30.9\n', 'IIB,09.01.22,24.6\n', 'IIB,10.01.22,14\n', 'IIB,11.01.22,39.4\n', 'IIB,12.01.22,41.9\n', 'IIB,13.01.22,43\n', 'IIB,14.01.22,34.9\n', 'IIB,15.01.22,18.5\n', 'IIB,16.01.22,28.9\n', 'IIB,17.01.22,29.1\n', 'IIB,18.01.22,38.3\n', 'IIB,19.01.22,21\n', 'IIB,20.01.22,4.3\n', 'IIB,21.01.22,15.6\n', 'IIB,22.01.22,24.4\n', 'IIB,23.01.22,19.4\n', 'IIB,24.01.22,26.1\n', 'IIB,25.01.22,4.1\n', 'IIB,26.01.22,17.6\n', 'IIB,27.01.22,35.2\n', 'IIB,28.01.22,14.9\n', 'IIB,29.01.22,29.8\n', 'IIB,30.01.22,24.8\n', 'IIB,31.01.22,55\n', 'IIB,01.02.22,37.9\n', 'IIB,02.02.22,12.9\n', 'IIB,03.02.22,11.5\n', 'IIB,04.02.22,16\n', 'IIB,05.02.22,28.4\n', 'IIB,06.02.22,36.7\n', 'IIB,07.02.22,22.6\n', 'IIB,08.02.22,41.5\n', 'IIB,09.02.22,39.4\n', 'IIB,10.02.22,38.7\n', 'IIB,11.02.22,39.9\n', 'IIB,12.02.22,13.8\n', 'IIB,13.02.22,12.2\n', 'I

**Step 3.** Build a dictionary mapping each month to the number of observations in it.

You do not know in advance which months are present, so the dictionary has to grow as you go.
Keep skipping the days with no reading: a day the sensor did not record is not an observation, so
October and November should come out short of their full length.

In [89]:
# Your solution here
# Hint: start with an empty dict, counts = {}
# Hint: counts.get(month, 0) returns 0 for a month you have not seen yet
# Hint: counts[month] = counts.get(month, 0) + 1

**Step 4.** Build a second dictionary mapping each month to its **mean** daily maximum
temperature in °C. Print it with each mean rounded to two decimals.

In [90]:
# Your solution here
# Hint: accumulate a running total per month in the same loop as the counts
# Hint: a second loop over the totals turns each total into a mean
# Hint: reuse your function from step 1 — do not repeat the conversion arithmetic

**Step 5.** Find the warmest month and print it with its mean temperature.

Do this with a loop and a comparison, keeping track of the best value seen so far, as you did for
the warmest day in Exercise 6. A dictionary is looked up by key, not by position, so there is no
`[0]` to start from: start from `None` instead.

In [91]:
# Your solution here
# Hint: initialise warmest_month = None and warmest_mean = None before the loop
# Hint: if warmest_mean is None, take this month; elif its mean is larger, take it instead

**Step 6.** Write your monthly summary to `_files/monthly_means.csv`, with the header
`month,n_obs,mean_temp_celsius` and one line per month, creating the `_files` folder first as you
did in 1.1. Then read the file back and print it, to confirm it says what you meant it to say.

In [92]:
# Your solution here
# Hint: out_path = Path("_files") / "monthly_means.csv" builds the path, and
# out_path.parent.mkdir(exist_ok=True) creates the _files folder if it is missing
# Hint: open with mode "w" inside a with block, and write the header first
# Hint: build each line with an f-string, and remember the "\n"
# Hint: f"{value:.2f}" keeps the file tidy

**Step 7 (optional).** This step uses `try`/`except` from the going-deeper box in the lecture. Real files contain junk, and you have already met one kind of it — the empty
temperature field. Suppose some lines are truncated as well, or hold a temperature that is not a
number at all. Modify your reading loop so that it skips any line that does not have three fields
or whose temperature will not convert to a number, counting how many lines it skipped.

Test it on this list before applying it to the real file:

```python
test_lines = [
    "IIB,01.01.22,30.2",
    "IIB,02.01.22",
    "IIB,03.01.22,notanumber",
    "IIB,04.01.22,33.4",
]
```

You should keep 2 and skip 2.

In [93]:
# Your solution here
# Hint: len(fields) != 3 catches the truncated line — use continue
# Hint: wrap the float() call in try/except ValueError to catch the bad number
# Hint: continue inside a for loop moves straight to the next line

**ℹ️ What you just did**

You wrote a reusable function, looped over a whole year of records, built dictionaries whose
keys you did not know in advance, aggregated by group, and wrote a derived dataset back to disk
— with a loop that survives malformed input. This is the shape of nearly every data-processing
script you will write. In subchapter 1.5, `pandas` will do the grouping in one line; the
reasoning will be exactly the same.

## Exercise 17: The solar system

<img src="https://raw.githubusercontent.com/gse-unil/2026_MLEES_book/main/part-I/_static/nasa-solar-system-montage.jpg" alt="The eight planets and the Moon photographed by spacecraft, arranged diagonally against black" width="100%">

<em>The eight planets and the Moon, as photographed by Mariner 10, Magellan, Galileo, Mars Global
Surveyor, Cassini and the Voyager spacecraft. The inner bodies are roughly to scale with each other, as are the outer ones, but the
two groups are not to scale with one another. Image
<a href="https://www.jpl.nasa.gov/images/pia03153-solar-system-montage-high-resolution-2001-version/">NASA/JPL (PIA03153)</a>, public domain.</em>


This is the long exercise for this subchapter, and the only one in it whose data is not
environmental. Eight planets and their masses are few enough to type by hand, and familiar enough
that a wrong answer is easy to spot, which makes them a good first test of lists, dictionaries and
functions together.

Work through sections A to C in order: section C writes functions against the dictionary you
build in section B. You will pick the exercise up again in 1.7, once classes are available, and
turn the dictionary into a `Planet` type.

The masses come from NASA's
[planetary fact sheet](https://nssdc.gsfc.nasa.gov/planetary/factsheet/). Use units of
10^24 kg throughout, so that Earth is 5.97 and Jupiter is 1898 — the variable names below carry
that unit, which is what makes a bare number like `1898` readable three cells later.

### A: lists and loops

**Q1) Create a list with the names of every planet in the solar system, in order.**

Name it `planets` and use lowercase names, so it starts `["mercury", "venus", ...]`.

In [116]:
# Q1: create your list here

planets = ["mercury", "venus", "earth", "mars", "jupiter", "saturn", "uranus", "neptune"]
print("planets", planets)

planets ['mercury', 'venus', 'earth', 'mars', 'jupiter', 'saturn', 'uranus', 'neptune']


**Q2) Have Python tell you how many planets there are by examining your list.**

You should get 8. Do not type the number — read it off the list.

In [117]:
# Q2: count the planets
# Hint: len() reads the length off the list itself
print("how many planets are there?", len(planets))

how many planets are there? 8


**Q3) Use slicing to display the first four planets — the rocky ones.**

In [118]:
# Q3: the rocky planets
print("first four planets:", planets[:4])

first four planets: ['mercury', 'venus', 'earth', 'mars']


**Q4) Iterate through your planets and print the planet name only if it ends in "s".**

You should see `venus`, `mars` and `uranus`, and nothing else.

Hint: the last letter of a name is `planet[-1]`. Use `continue` for the names that do not match,
so the loop body says explicitly what it skips.

In [119]:
# Q4: print the planets whose name ends in s

for planet in planets:
  if planet[-1] != "s":
    continue
  print(planet)

venus
mars
uranus


### B: dictionary

**Q5) Create a dictionary that maps each planet name to its mass.**

Call it `planet_masses_1e24kg`, and read the values off the
[NASA fact sheet](https://nssdc.gsfc.nasa.gov/planetary/factsheet/) in units of 10^24 kg. Every
value has to be in that one unit: entered as 1.898 (in 10^27 kg), Jupiter would come out 1000
times too light in every comparison and ratio below.

In [131]:
# Q5: build the dictionary here

planet_masses_1e24kg = {
    "mercury": 0.330,
    "venus": 4.87,
    "earth": 5.97,
    "mars": 0.642,
    "jupiter": 1898,
    "saturn": 568,
    "uranus": 86.8,
    "neptune": 102,
}

planet_masses_1e24kg.items()

dict_items([('mercury', 0.33), ('venus', 4.87), ('earth', 5.97), ('mars', 0.642), ('jupiter', 1898), ('saturn', 568), ('uranus', 86.8), ('neptune', 102)])

**Q6) Use your dictionary to look up Earth's mass.**

In [122]:
# Q6: look up Earth
print(planet_masses_1e24kg["earth"])

# Trying using get
planet_masses_1e24kg.get("earth")

5.97


5.97

**Q7) Loop through the dictionary and build a list of every planet heavier than
100 x 10^24 kg.**

You should end up with `["jupiter", "saturn", "neptune"]` — uranus, at 86.8, does not make it.

Hint: `.items()` gives you the name and the mass together on each pass, as in 1.2's lecture.

In [129]:
# Q7: the planets heavier than 100e24 kg

# Build a list for heavier planets
heavier_planets = [] # collector for list

for planet, planet_mass in planet_masses_1e24kg.items():
  if planet_mass > 100:
    heavier_planets.append(planet) # add the answer to the collector list

# Print collector list
print("list of planets heavier than 100x10^24kg:", heavier_planets)

list of planets heavier than 100x10^24kg: ['jupiter', 'saturn', 'neptune']


**Q8) Add Pluto to your dictionary.**

Its mass is 0.0130 x 10^24 kg. Print the names afterwards, looping with `.items()`, to confirm it
landed.

Pluto has not been a planet since 2006. Adding it still takes one assignment, and none of the code
that reads the other eight entries has to change.

In [140]:
# Q8: add Pluto, then print the names
planet_masses_1e24kg["pluto"] = 0.0130
for planet, planet_mass in planet_masses_1e24kg.items():
    print(planet) # pluto has been added

mercury
venus
earth
mars
jupiter
saturn
uranus
neptune
pluto


### C: functions

**Q9) Write a function that converts a planetary mass to Earth masses.**

Jupiter is about 1898 x 10^24 kg, which is about 318 Earth masses:

$$M_{\text{Jupiter}} \approx 1898 \times 10^{24}\,\text{kg} \approx 318\,M_{\text{Earth}}$$

Call it `to_earth_masses(mass_1e24kg)` and check it against Jupiter — you should get
317.92294807370183.

In [143]:
# Q9: write to_earth_masses here, then test it on Jupiter
def to_earth_masses(mass_1e24kg):
  """Return the mass in Earth masses"""
  for planet in planet_masses_1e24kg:
    return mass_1e24kg / planet_masses_1e24kg["earth"]

print("call to check against jupiter:", to_earth_masses(planet_masses_1e24kg["jupiter"])) # 317.92294807370183.

call to check against jupiter: 317.92294807370183


**Q10) Now write a single function that converts to either Earth masses or Jupiter
masses, depending on a keyword argument.**

Call it `to_planet_masses(mass_1e24kg, reference="jupiter")`. Converting Jupiter's own mass should
give 317.923 with `reference="earth"` and exactly 1.0 with `reference="jupiter"`.

With the default argument, one function covers both conversions. Jupiter in Jupiter masses has
to come out as exactly 1.0; any other value means the function divides by the wrong mass.

In [156]:
# Q10: write to_planet_masses here, then test both references on Jupiter
def to_planet_masses(mass_1e24kg, reference="jupiter"):
  for planet in planet_masses_1e24kg:
    """Convert a mass to either jupiter/earth masses"""
    if reference == "jupiter":
      return mass_1e24kg / planet_masses_1e24kg["jupiter"]
    elif reference == "earth":
      return mass_1e24kg / planet_masses_1e24kg["earth"]

# Convert jupiter own mass
print("covert to jupiter own mass:", to_planet_masses(1898)) # auto assign to jupiter?
print("convert to jupiter own mass:", to_planet_masses(1898, reference="jupiter"))

# Convert to earth masses
print("convert to earth own mass:", to_planet_masses(1898, reference="earth")) # assign to earth masses

covert to jupiter own mass: 1.0
convert to jupiter own mass: 1.0
convert to earth own mass: 317.92294807370183


**Q11) Write a function that returns two values.**

`mass_in_two_units(mass_1e24kg, reference)` should return the mass in Earth masses *and* in masses
of whatever reference planet you pass. Test it on Jupiter with `reference="mars"`.

Hint: `return a, b` returns a tuple, and `first, second = f(...)` unpacks it — both from 1.2's
lecture.

In [159]:
# Q11: write mass_in_two_units here, then test it on Jupiter against mars
def mass_in_two_units(mass_1e24kg, reference):
  for planet in planet_masses_1e24kg:
    earth_conversion = mass_1e24kg / planet_masses_1e24kg["earth"]
    in_reference = mass_1e24kg / planet_masses_1e24kg[reference]
    """Convert a mass into two units"""
    return earth_conversion, in_reference

# print in two masses
print("test on jupiter for earth and mars:", mass_in_two_units(1898, "mars"))

# try unpack
earth_conversion, in_reference = mass_in_two_units(1898, "mars")
print("earth conversion:", earth_conversion, "in reference:", in_reference)

test on jupiter for earth and mars: (317.92294807370183, 2956.386292834891)
earth conversion: 317.92294807370183 in reference: 2956.386292834891


**Bonus.** Convert Neptune's mass to Jupiter masses with your Q10 function. Then pass that result
to your Q11 function with `reference="jupiter"`, and compare the second value it returns with the
Q10 result. Are they the same? If not, work out which function is being given a number in the
wrong unit, and what it should have been given instead — this is the mistake the `_1e24kg` suffix
exists to prevent.

In [162]:
# Bonus: Neptune, out and back

# Q10: write to_planet_masses here, then test both references on Jupiter
in_jupiters = to_planet_masses(1898)         # 102, in 10^24 kg

# Q10: Neptune in Jupiter masses
jupiters = to_planet_masses(102, reference="jupiter")  # 102 / 1898 ≈ 0.0537

# Q11: pass that RESULT to the two-unit function
earths, in_jupiters = mass_in_two_units(in_jupiters, reference="jupiter")
print(in_jupiters, jupiters) # same result!

0.0005268703898840885 0.05374077976817703


**ℹ️ What you just did**

You built a list and read its length rather than counting it, sliced a subset out of it, and
looped over it with a condition. You built a dictionary keyed by name, looked a value up, filtered
it into a new list, and extended it. Then you wrote three functions over that dictionary: one
fixed, one with a default argument that collapses two functions into one, and one returning a pair.

That is the whole of this subchapter applied to one dataset. In 1.7 you will come back and replace
the dictionary with a class, so that a planet's name and mass travel together as one object
instead of as a key and a value that happen to line up.